frontend/src/routes/Upload.tsx

```diff
@@ -71,6 +77,26 @@ const Upload: React.FC = () => {
     );
 };
 
+let controller = new AbortController();
```

Comment:
Why `useImperativeHandle`? The way I see it, the `Upload` component can simply create an `AbortController` (probably store it as a ref) and then pass it through to the cancel button and the `uploadTracks` function.

Reply:
Or that. This is one of the situations, though, where I find React's "top-down" flow a bit unintuitive. The cancel button is the thing that should be controlling the abort controller, IMO. 🤷

Reply:
I tried doing this with `useImperativeHandle` yesterday after looking at some examples, but couldn't quite get it to work. I may have misunderstood what you meant by "exposing" it via the cancel button. Did you mean the controller should be instanciated inside the button? 
Since the `AbortController` also needs to be called when leaving the route, I'm not sure how this would be done if it is only controlled by the cancel button, apart from simulating a click or sth like that.

Reply:
@JulianKniephoff Why would the cancel button control it? There are two things that can trigger an abort: the cancel button and navigating away from the route. So I think just creating the `AbortController` in the `Upload` component is the more natural place.

Reply:
@owi92 yes that's what I meant: `CancelButton` would instantiate it (probably in a `useEffect` callback), and then -- using `forwardRef` and `useImperativeHandle`, you can get access to it from the outside. That is, you can make something like this work:

```typescript
// Let's say we are in `UploadState`, which currently renders `CancelButton`:
// [...]
const abortControllerRef = useRef<AbortController>(null);
// [...]
return <Whatever>
    {/* [...] */}
    <CancelButton ref={abortControllerRef} /* [...] */ />
    {/* [...] */}
</Whatever>;
```

By passing `abortControllerRef` as `ref` to `CancelButton` (when that component contains the proper incantations using `forwardRef` and `useImperativeHandle`) that ref can be "bound" to the abort controller living inside of `CancelButton`. That is, you can now, in some callback in `UploadState` for example, access that abort controller as `abortControllerRef.current`. That's what I mean by "expose."

One problem I just noticed is that you basically need to do that twice, because you need the controller in `Upload`, not in `UploadState`. That's a bit unfortunate. But it should work.

You can try to make it work (and consult me if you get stuck) still, but in light of the above I'd also be fine with what @LukasKalbertodt suggests: Just create the controller in `Upload` and pass it around to wherever it is needed, or put it in a (React) context.

By the way the fact that we also want to abort when we leave the page isn't actually the problem here: The easiest way to solve **that** would actually be to let `CancelButton` itself listen for the appropriate event. However, we still need the abort controller in functions controlled and called by the `Upload` component, so that doesn't really help us. 🤷